eth-rpc/serverdispatch.nim

```diff
@@ -0,0 +1,295 @@
+import json, tables, strutils, options, macros
+import asyncdispatch2
+import jsonmarshal
+
+export json, jsonmarshal
+
+type
+  RpcJsonError* = enum
+    rjeInvalidJson, rjeVersionError, rjeNoMethod, rjeNoId
+
+  RpcJsonErrorContainer* = tuple[err: RpcJsonError, msg: string]
+
+  RpcProc* = proc (params: JsonNode): Future[JsonNode]
+
+  RpcServer* = ref object of RootRef
+    server*: StreamServer
+    procs*: TableRef[string, RpcProc]
+
+  RpcProcError* = ref object of Exception
+    code*: int
+    data*: JsonNode
+
+const
+  JSON_PARSE_ERROR* = -32700
+  INVALID_REQUEST* = -32600
+  METHOD_NOT_FOUND* = -32601
+  INVALID_PARAMS* = -32602
+  INTERNAL_ERROR* = -32603
+  SERVER_ERROR* = -32000
+
+  jsonErrorMessages*: array[RpcJsonError, (int, string)] =
+    [
+      (JSON_PARSE_ERROR, "Invalid JSON"),
+      (INVALID_REQUEST, "JSON 2.0 required"),
+      (INVALID_REQUEST, "No method requested"),
+      (INVALID_REQUEST, "No id specified")
+    ]
+
+when not defined(release):
+  template ifDebug*(actions: untyped): untyped =
+    actions
+else:
+  template ifDebug*(actions: untyped): untyped = discard
+
+template jsonValid*(jsonString: string, node: var JsonNode): (bool, string) =
+  var
+    valid = true
+    msg = ""
+  try: node = parseJson(line)
+  except:
+    valid = false
+    msg = getCurrentExceptionMsg()
+  (valid, msg)
+
+proc checkJsonErrors*(line: string,
+                      node: var JsonNode): Option[RpcJsonErrorContainer] =
+  ## Tries parsing line into node, if successful checks required fields
+  ## Returns: error state or none
+  let res = jsonValid(line, node)
+  if not res[0]:
+    return some((rjeInvalidJson, res[1]))
+  if not node.hasKey("jsonrpc"):
+    return some((rjeVersionError, ""))
+  if not node.hasKey("method"):
+    return some((rjeNoMethod, ""))
+  if not node.hasKey("id"):
+    return some((rjeNoId, ""))
+  return none(RpcJsonErrorContainer)
+
+proc wrapReply*(id: JsonNode, value: JsonNode, error: JsonNode): JsonNode =
+  return %{"jsonrpc": %"2.0", "result": value, "error": error, "id": id}
+
+proc sendError*(client: StreamTransport, code: int, msg: string,
+                id: JsonNode, data: JsonNode = newJNull()) {.async.} =
+  ## Send error message to client
+  let error = %{"code": %(code), "message": %msg, "data": data}
+  ifDebug: echo "Send error json: ",
+                wrapReply(newJNull(), error, id).pretty & "\c\l"
```

Comment:
A note about performance:

Such string concatenations are inefficient with the current Nim, because they lead to copying all of the concatenated buffers. It's more efficient to replace the concatenation with extra arguments to `echo`:

``` nim
echo "foo", bar, "\c\l"
```

There is a disclaimer though: future version of Nim will be able to optimize the copy away by stealing the buffer of the temporary strings involved (similar to C++'s rvalue references).

Reply:
These will be replaced by `nim-chronicle` messages soon.